src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_keychain.c

```diff
@@ -224,6 +224,100 @@ static OSStatus DeleteInKeychain(CFTypeRef needle, SecKeychainRef haystack)
     return status;
 }
 
+static bool IsCertInKeychain(CFTypeRef needle, SecKeychainRef haystack)
+{
+    CFMutableDictionaryRef query = CFDictionaryCreateMutable(
+        kCFAllocatorDefault, 0, &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks);
+
+    if (query == NULL)
+        return errSecAllocate;
+
+    const void* constHaystack = haystack;
+    CFTypeRef result = NULL;
+    CFArrayRef searchList = CFArrayCreate(NULL, &constHaystack, 1, &kCFTypeArrayCallBacks);
+
+    if (searchList == NULL)
+    {
+        CFRelease(query);
+        return errSecAllocate;
+    }
+
+    CFArrayRef itemMatch = CFArrayCreate(NULL, (const void**)(&needle), 1, &kCFTypeArrayCallBacks);
+
+    if (itemMatch == NULL)
+    {
+        CFRelease(searchList);
+        CFRelease(query);
+        return errSecAllocate;
+    }
+
+    CFDictionarySetValue(query, kSecReturnRef, kCFBooleanTrue);
+    CFDictionarySetValue(query, kSecMatchSearchList, searchList);
+    CFDictionarySetValue(query, kSecMatchItemList, itemMatch);
+    CFDictionarySetValue(query, kSecClass, kSecClassCertificate);
+    OSStatus status = SecItemCopyMatching(query, &result);
+
+    bool ret = true;
+
+    if (status == errSecItemNotFound)
+    {
+        CFDictionarySetValue(query, kSecClass, kSecClassIdentity);
+        status = SecItemCopyMatching(query, &result);
+
+        if (status == errSecItemNotFound)
+        {
+            ret = false;
```

Comment:
I don't think this block can be missed.  If an identity could be found (both a cert and a private key exist) then the cert should already have been found...

Usually the double-dips go the other way (identity then certificate) because they need the best thing possible.

Reply:
I'm not entirely sure what you mean by this. Are you setting that if the first time we call SecItemCopyMatching and it didn't find anything, that there is no need to do the secod SecItemCopyMatching call?

Reply:
Yes, that's what @bartonjs suggested and it seems be reasonable.